### PR TITLE
[cmake] add support for cc2650, qpg6095, samr21

### PR DESCRIPTION
**Pull request by FiveDimensions**

### CODE REVIEW

File: examples/platforms/cc2650/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+
+list(APPEND OT_PLATFORM_DEFINES
+    "OPENTHREAD_CORE_CONFIG_PLATFORM_CHECK_FILE=\"openthread-core-cc2650-config-check.h\""
+)
+set(OT_PLATFORM_DEFINES ${OT_PLATFORM_DEFINES} PARENT_SCOPE)
+
+list(APPEND OT_PRIVATE_DEFINES "MBEDTLS_USER_CONFIG_FILE=\"cc2650-mbedtls-config.h\"")
+set(OT_PRIVATE_DEFINES ${OT_PRIVATE_DEFINES} PARENT_SCOPE)
+list(APPEND OT_PUBLIC_INCLUDES
+    "${PROJECT_SOURCE_DIR}/examples/platforms/cc2650/crypto"
```
Comment (gjc13) [inline examples/platforms/cc2650/CMakeLists.txt:44]: ${CMAKE_CURRRENT_DIRECTORY}/crypto

File: examples/platforms/cc2650/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+    uart.c
+    $<TARGET_OBJECTS:openthread-platform-utils>
+)
+
+set_property(TARGET openthread-cc2650 PROPERTY C_STANDARD 99)
+
+target_link_libraries(openthread-cc2650 PRIVATE
+    mbedtls
+    openthread-platform-utils
+    "${PROJECT_SOURCE_DIR}/third_party/ti/devices/cc26x0/driverlib/bin/gcc/driverlib.a"
```
Comment (gjc13) [inline examples/platforms/cc2650/CMakeLists.txt:73]: Add a library under third_party/ti/devices depend on it here for link libraries and include dependencies.

File: examples/platforms/cc2650/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+)
+target_link_options(openthread-cc2650 PUBLIC
+    -T${PROJECT_SOURCE_DIR}/third_party/ti/devices/cc26x0/linker_files/cc26x0f128.lds
```
Comment (gjc13) [inline examples/platforms/cc2650/CMakeLists.txt:76]: Is it possible to specify ld script in the ti library?

File: examples/platforms/cc2650/arm-none-eabi.cmake [hunk 4]
```diff
@@ -0,0 +1,52 @@
+#
```
Comment (gjc13) [inline examples/platforms/cc2650/arm-none-eabi.cmake:1]: (Can be resolved afterwards)
Duplicate code in ${PLATFORM_NAME}/arm-none-eabi.cmake

File: examples/platforms/qpg6095/CMakeLists.txt [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+    platform.c
+    radio.c
+    settings.cpp
+    uart.c
+)
+
+set_property(TARGET openthread-qpg6095 PROPERTY C_STANDARD 99)
+
+target_link_libraries(openthread-qpg6095
+    PRIVATE
```
Comment (gjc13) [inline examples/platforms/qpg6095/CMakeLists.txt:64]: The same, add a target under `third_party/Qorvo/qpg6095`.

File: examples/platforms/qpg6095/CMakeLists.txt [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+    PUBLIC
+        ${OT_PLATFORM_DEFINES}
+    PRIVATE
+        ${OT_PRIVATE_DEFINES}
+)
+
+target_compile_options(openthread-qpg6095 PRIVATE
+    ${OT_CFLAGS}
+    -lrt
+    -lpthread
```
Comment (gjc13) [inline examples/platforms/qpg6095/CMakeLists.txt:85]: Why pthread is required for OpenThread?

File: examples/platforms/samr21/CMakeLists.txt [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+    cxx_helpers.c
+    diag.c
+    entropy.c
+    flash.c
+    logging.c
+    misc.c
+    radio.c
+    system.c
+    uart.c
+    ${PROJECT_SOURCE_DIR}/third_party/microchip/asf/common/utils/interrupt/interrupt_sam_nvic.c
```
Comment (gjc13) [inline examples/platforms/samr21/CMakeLists.txt:62]: The same, add a target under `third_party/microchip/asf`.

File: examples/platforms/qpg6095/CMakeLists.txt (hunk 6 above)
Comment (Irving-cl) [inline examples/platforms/qpg6095/CMakeLists.txt:85]: Shall we put these into `target_link_libraries`?

File: script/check-arm-build-cmake [hunk 8]
```diff
@@ -104,11 +104,83 @@ build_cc2652()
...
+    ninja
+    cd ..
+}
+
 main()
 {
     build_cc1352
     build_cc2538
     build_cc2652
+    build_cc2650
```
Comment (Irving-cl) [inline script/check-arm-build-cmake:181]: Shall we put `cc2650` before `cc2652` to make it in order?

File: examples/platforms/samr21/CMakeLists.txt (hunk 7 above)
Comment (FiveDimensions) [reply examples/platforms/samr21/CMakeLists.txt:62]: asf is not included in openthread repo and download temporarily when checking build

File: script/check-arm-build-cmake (hunk 8 above)
Comment (FiveDimensions) [reply script/check-arm-build-cmake:181]: sure

File: examples/platforms/cc2650/CMakeLists.txt (hunk 1 above)
Comment (FiveDimensions) [reply examples/platforms/cc2650/CMakeLists.txt:44]: You mean ${CMAKE_CURRENT_SOURCE_DIR}/crypto?

File: examples/platforms/qpg6095/CMakeLists.txt (hunk 5 above)
Comment (FiveDimensions) [reply examples/platforms/qpg6095/CMakeLists.txt:64]: It is an existing library and no need to build, should we need to add a target?

File: examples/platforms/cc2650/CMakeLists.txt (hunk 2 above)
Comment (FiveDimensions) [reply examples/platforms/cc2650/CMakeLists.txt:73]: It is a existing library, why should we add a library under third_party/ti/devices?

File: examples/platforms/qpg6095/CMakeLists.txt (hunk 6 above)
Comment (FiveDimensions) [reply examples/platforms/qpg6095/CMakeLists.txt:85]: inherited from makefile, not found either， delete

File: examples/platforms/qpg6095/CMakeLists.txt [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+target_compile_options(openthread-qpg6095
+    PRIVATE
+        ${OT_CFLAGS}
+)
+
+target_include_directories(openthread-qpg6095
+    PRIVATE
+        ${OT_PUBLIC_INCLUDES}
+        ${OT_PRIVATE_INCLUDES}
+        ${PROJECT_SOURCE_DIR}/src/core
```
Comment (gjc13) [inline examples/platforms/qpg6095/CMakeLists.txt:91]: The same.

File: examples/platforms/qpg6095/CMakeLists.txt [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+        ${PROJECT_SOURCE_DIR}/examples/platforms
+        ${PROJECT_SOURCE_DIR}/third_party/mbedtls/repo/include
```
Comment (gjc13) [inline examples/platforms/qpg6095/CMakeLists.txt:93]: The same.

File: examples/platforms/cc2650/CMakeLists.txt [hunk 11]
```diff
@@ -0,0 +1,100 @@
...
+        ${OT_PRIVATE_DEFINES}
+)
+
+target_compile_options(openthread-cc2650 PRIVATE ${OT_CFLAGS})
+
+target_include_directories(openthread-cc2650
+    PRIVATE
+        ${OT_PUBLIC_INCLUDES}
+        ${OT_PRIVATE_INCLUDES}
+        ${PROJECT_SOURCE_DIR}/src/core
```
Comment (gjc13) [inline examples/platforms/cc2650/CMakeLists.txt:96]: Use OT_PRIVATE_INCLUDE instead?

File: examples/platforms/cc2650/CMakeLists.txt [hunk 12]
```diff
@@ -0,0 +1,100 @@
...
+        ${PROJECT_SOURCE_DIR}/examples/platforms
+        ${PROJECT_SOURCE_DIR}/third_party/mbedtls/repo/include
```
Comment (gjc13) [inline examples/platforms/cc2650/CMakeLists.txt:98]: Can we add target_link_libraries(PUBLIC mbedtls) here?

File: examples/platforms/samr21/CMakeLists.txt [hunk 13]
```diff
@@ -0,0 +1,99 @@
...
+target_compile_options(openthread-samr21
+    PRIVATE
+        ${OT_CFLAGS}
+)
+
+target_include_directories(openthread-samr21
+    PRIVATE
+        ${OT_PUBLIC_INCLUDES}
+        ${OT_PRIVATE_INCLUDES}
+        ${PROJECT_SOURCE_DIR}/src/core
```
Comment (gjc13) [inline examples/platforms/samr21/CMakeLists.txt:95]: The same.

File: examples/platforms/samr21/CMakeLists.txt [hunk 14]
```diff
@@ -0,0 +1,99 @@
...
+        ${PROJECT_SOURCE_DIR}/examples/platforms
+        ${PROJECT_SOURCE_DIR}/third_party/mbedtls/repo/include
```
Comment (gjc13) [inline examples/platforms/samr21/CMakeLists.txt:97]: The same.

File: examples/platforms/cc2650/CMakeLists.txt (hunk 12 above)
Comment (FiveDimensions) [reply examples/platforms/cc2650/CMakeLists.txt:98]: we can make it in another pr for those related platforms.

File: examples/platforms/cc2650/CMakeLists.txt (hunk 11 above)
Comment (gjc13) [reply examples/platforms/cc2650/CMakeLists.txt:96]: ${PROJECT_SOURCE_DIR}/src/core turned out to be not in ${OT_PRIVATE_INCLUDES}.

File: examples/platforms/qpg6095/CMakeLists.txt [hunk 15]
```diff
@@ -0,0 +1,95 @@
...
+    "OPENTHREAD_CORE_CONFIG_PLATFORM_CHECK_FILE=\"openthread-core-qpg6095-config-check.h\""
+)
+set(OT_PLATFORM_DEFINES ${OT_PLATFORM_DEFINES} PARENT_SCOPE)
+
+list(APPEND OT_PRIVATE_DEFINES "MBEDTLS_USER_CONFIG_FILE=\"qpg6095-mbedtls-config.h\"")
+list(APPEND OT_PUBLIC_INCLUDES
+    "${CMAKE_CURRENT_SOURCE_DIR}/crypto"
+)
+set(OT_PUBLIC_INCLUDES ${OT_PUBLIC_INCLUDES} PARENT_SCOPE)
+string(REPLACE "-pedantic-errors" "" OT_CFLAGS "${OT_CFLAGS}")
```
Comment (librasungirl) [inline examples/platforms/qpg6095/CMakeLists.txt:46]: may replace when there is via `if` like in https://cmake.org/pipermail/cmake/2008-February/019614.html ?

File: third_party/ti/CMakeLists.txt [hunk 16]
```diff
@@ -0,0 +1,41 @@
...
+#  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+#  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+#  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+#  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+#  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+#  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+#  POSSIBILITY OF SUCH DAMAGE.
+#
+
+if(OT_PLATFORM STREQUAL "cc2650")
```
Comment (jwhui) [inline third_party/ti/CMakeLists.txt:29]: Does this also apply to cc2652?

Comment (FiveDimensions) [reply third_party/ti/CMakeLists.txt:29]: no, cc2652 use cc13x2_cc26x2/driverlib/bin/gcc/driverlib.a and will update after this

**Pull request merged**